src/Json/Internal/String.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+
+module Json.Internal.String
+  (
+  ) where
+
+import Prelude hiding (length)
+
+import Data.Bits (complement, (.&.), (.|.), xor)
+import Data.Bytes (Bytes)
+import Data.Bytes.Types (Bytes(..))
+import Data.Primitive.ByteArray (indexByteArray)
+import Data.Word (Word64)
+
+import qualified Data.Bytes as Bytes
+
+
+-- P.uneffectful
+
+advance :: Bytes -> Result
+{-# inline advance #-}
+advance bs0 = loop bs0 CanMemcpy
+  where
+  loop bs canMemcpy = case advAlign bs canMemcpy of
+    Continue canMemcpy' bs' -> case advWords bs' canMemcpy' of
+      Continue canMemcpy'' bs'' -> case advChar bs'' canMemcpy'' of
+        Continue canMemcpy''' bs''' -> loop bs''' canMemcpy'''
+        res -> res
+      res -> res
+    res -> res
+
+
+data CanMemcpy = CanMemcpy | NoMemcpy
+data Result
```

Comment:
Put bangs on all the fields except for `CanMemcpy`, which doesn't need it. Explicitly `UNPACK` the `Bytes` fields as well.